File: go/polly/DescribeVoices/DescribeVoices.go
// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: MIT-0
// snippet-start:[polly.go.describe_voices]
package main

// snippet-start:[polly.go.describe_voices.imports]
import (
    "fmt"

    "github.com/aws/aws-sdk-go/aws"
    "github.com/aws/aws-sdk-go/aws/session"
    "github.com/aws/aws-sdk-go/service/polly"
    "github.com/aws/aws-sdk-go/service/polly/pollyiface"
)
// snippet-end:[polly.go.describe_voices.imports]

// GetVoices retrieves a list of the voices for US English.
// Inputs:
//     svc is a Polly service client
// Output:
//     If success, the list of voices and nil
//     Otherwise, nil and an error from the call to DescribeVoices
func GetVoices(svc pollyiface.PollyAPI) (*polly.DescribeVoicesOutput, error) {
    // snippet-start:[polly.go.describe_voices.call]
    resp, err := svc.DescribeVoices(&polly.DescribeVoicesInput{
        LanguageCode: aws.String("en-US"),
    })
    // snippet-end:[polly.go.describe_voices.call]
    return resp, err
}

func main() {
    // snippet-start:[polly.go.describe_voices.session]
    sess := session.Must(session.NewSessionWithOptions(session.Options{
        SharedConfigState: session.SharedConfigEnable,
    }))

    svc := polly.New(sess)
    // snippet-end:[polly.go.describe_voices.session]
    
    resp, err := GetVoices(svc)
    if err != nil {
        fmt.Println("Got error calling DescribeVoices:")
        fmt.Print(err.Error())
        return
    }

    // snippet-start:[polly.go.describe_voices.display]
    for _, v := range resp.Voices {
        fmt.Println("Name:   " + *v.Name)
        fmt.Println("Gender: " + *v.Gender)
        fmt.Println("")
    }
    // snippet-end:[polly.go.describe_voices.display]
}
// snippet-end:[polly.go.describe_voices]

File: rust_dev_preview/examples/polly/src/bin/describe-voices.rs
/*
 * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 * SPDX-License-Identifier: Apache-2.0.
 */

#![allow(clippy::result_large_err)]

use aws_config::meta::region::RegionProviderChain;
use aws_sdk_polly::{config::Region, meta::PKG_VERSION, Client, Error};
use clap::Parser;

#[derive(Debug, Parser)]
struct Opt {
    /// The AWS Region.
    #[structopt(short, long)]
    region: Option<String>,

    /// Whether to isplay additional information.
    #[structopt(short, long)]
    verbose: bool,
}

// Lists the available voices.
// snippet-start:[polly.rust.describe-voices]
async fn list_voices(client: &Client) -> Result<(), Error> {
    let resp = client.describe_voices().send().await?;

    println!("Voices:");

    let voices = resp.voices();
    for voice in voices {
        println!("  Name:     {}", voice.name().unwrap_or("No name!"));
        println!(
            "  Language: {}",
            voice.language_name().unwrap_or("No language!")
        );

        println!();
    }

    println!("Found {} voices", voices.len());

    Ok(())
}
// snippet-end:[polly.rust.describe-voices]

/// Displays a list of the voices in the Region.
/// # Arguments
///
/// * `[-r REGION]` - The Region in which the client is created.
///    If not supplied, uses the value of the **AWS_REGION** environment variable.
///    If the environment variable is not set, defaults to **us-west-2**.
/// * `[-v]` - Whether to display additional information.
#[tokio::main]
async fn main() -> Result<(), Error> {
    tracing_subscriber::fmt::init();

    let Opt { region, verbose } = Opt::parse();

    let region_provider = RegionProviderChain::first_try(region.map(Region::new))
        .or_default_provider()
        .or_else(Region::new("us-west-2"));
    println!();

    if verbose {
        println!("Polly client version: {}", PKG_VERSION);
        println!(
            "Region:               {}",
            region_provider.region().await.unwrap().as_ref()
        );
        println!();
    }

    let shared_config = aws_config::from_env().region(region_provider).load().await;
    let client = Client::new(&shared_config);

    list_voices(&client).await
}
